### ISSUE TITLE
built in input state tracking

### ISSUE THREAD

**Issue by adolson** [godotengine/godot#1766]
is_action_pressed() should be able to return an integer corresponding to the input state, without breaking backwards-compatibility. For example, if 0 was NOT_PRESSED, 1 is JUST_PRESSED, 2 is PRESSED, and 3 is JUST_RELEASED. (You'd get the 3 only for one frame I'd imagine, when releasing the button. It shouldn't actually affect anything that is just treating the function as boolean).

@ndee85 has a GDScript which does something similar to this, but it'd be better if it were built in and standardized.

**Comment by punto-** [comment]
In my opinion, this is simple enough to do in script that I never needed to
have it as an engine feature. Regardless of that, I don't think it's
necessary to maintain backwards compatibility, it could be a new method on
Input, and in that case you'd avoid the case of just_released being true
(which nowadays would return false, so it does break it technically).

**Comment by reduz** [comment]
if it's a common use function, I don't mind adding it to Input , but was
this not added long ago?

**Comment by reduz** [comment]
ah i guess not

**Comment by shackra** [comment]
May can be this applied to [actions](https://github.com/okamstudio/godot/wiki/class_inputeventaction) as well?

**Comment by OlexiyKravchuk** [comment]
I propose to such names for new functions

is_action_push () / or / is_action_press () #. It triggered only once when the button is down
is_action_release () #. It is triggered only once when the button is up

is_action_hold () #. works constantly all the time, ie as well as is_action_pressed () exists now

**Comment by Biliogadafr** [comment]
I have some questions about behaviour of this feature. 
   1) How it should work inside _update() and _fixed_update() ? looks like If two _update() happens per one _fixed_update(), then 'Input' should maintain this "JUST_x" states seaparately for each kind of 'update' to avoid for example double "JUST_PRESSED" in 2 frames of _update or miss of event in _fixed_update  
   2) Is it possible to have situation when JUST_PRESSED and JUST_RELEASED happens in the same frame (maybe because of occasional lag that made frame longer)? In this case JUST_RELEASED probably should be moved to next frame somehow or the whole press ignored...
    It sounds comfortable to use, but also it is a bit strange when events attempted to be used as flags.

**Comment by reduz** [comment]
yes I thought about this, if you have process and fixed process, a function just_pressed would make no sense because the behaviour would be different. How do other game engines handle this?

**Comment by OlexiyKravchuk** [comment]
@reduz
it is just an idea ... I propose the concept of using the event history of the previous step in this form of flags ...

0) in the past is not pressed, and now not pressed; == not pressed;
1) Past performance is not pressed, but now pressed;
2) The in the past pressed, and now pressed; == pressed;
3), in the past pressed, but now not pressed;

**Comment by MarianoGnu** [comment]
Andreas Esau and myself did a gdscript for this behaviour but it has a several problem: if you check for key state twise in the same _proccess() call it will move from JUST_PRESSED to HOLD and from JUST_RELEASED to NOT_PRESSED.

To avoid this you need to make a global singleton ti track all your actions in _proccess() and _fixed_proccess() and store the states in separate variables.

This is my implementation, with the problem i mentioned:
https://gist.github.com/MarianoGnu/ed365c2a2a5eeb2e8b60

**Comment by razvanc87** [comment]
There's been quite some time since this was last commented on so... I'm putting it back into focus. I think there are basic things that should be supported in the engine :). Don't know how other engines implement this thing tho'... have no experience here...

**Comment by reduz** [comment]
Oh hey what about this? We can make the function work differently  depending on whether you are inside a fixed process loop or idle process loop...

**Comment by razvanc87** [comment]
Haha, I did something similar in GDScript, except I found that you have to distinguish between using it in `_input` vs `_process` or `_fixed_process`. I've improved on @ndee85's idea and commented everything so it's clear what's what... or so I hope:

```gdscript
extends CanvasLayer


var ActionState = preload("res://misc/scripts/action_state.gd")
var actions = {}


func _ready():
	self.set_process_input(true)
	for name in ["platform_fire", "platform_move"]:
		self.actions[name] = ActionState.new(name)



func _input(evt):
	print("platform_fire: ", self.actions["platform_fire"].get_state())
```

Hope this helps for the core implementation.

**Comment by razvanc87** [comment]
I further purpose these states: `IDLE`, `PRESSED`,  `HOLD`, `RELEASED` with some predefined time for checking the state. For example, 100ms to be in `PRESSED `state before moving to `HOLD` or `RELEASED` states. This would make it independent of fixed or idle processing. And it also makes sense for input processing.

**Comment by punto-** [comment]
that means you'll get "pressed" for more than 1 frame (at least 7 if it's
100ms), you want this to happen only for 1 frame. Why not just listen for
input events and do whatever you need when you see the event you want?

**Comment by razvanc87** [comment]
Well, that's the thing, the complication comes because instead of having pressed and released, you have pressed, hold and release events. If you want to check in both fixed and idle process for pressed (not hold)... Then how the heck do you do it?...

**Comment by punto-** [comment]
You don't have a "hold" event, you have pressed and released. You get those
events only once, when the button is pressed and released. In between you
get nothing (unless it's a keyboard, then you get echo events). the Input
class keeps track of the state of the key, but those events will tell you
when a key was pressed, you can do whatever you want with them..

**Comment by razvanc87** [comment]
Sorry, maybe I want clear enough. You get pressed and released in the `InputEvent` class (which is used in `_input `, no?), but I was talking about the `Input` class as well which can be used in `_process`, `_fixed_process ` and `_input ` and which only has pressed...

**Comment by razvanc87** [comment]
OK, so this is my final (working!) solution. This is what I'm talking about, when I was referring to using it in all processing functions. Now if only this can be included in the core engine with only one function call `.get_state()`... that would be ideal :).

This is `action_script.gd`:

```gdscript
# states
const IDLE = 0
const PRESSED = 1
const HOLD = 2
const RELEASED = 3
# private variables
var __state_prev = {"idle": false, "fixed": false, "input": false}  # the previous state
var __state = {"idle": false, "fixed": false, "input": false}  # the current state
# public variables
var name = ""  # the name of the action


func _init(name, input_processing=true):
	"""
	This class is for checking the state of the actions (from input map). It can check
	if the action is just pressed, pressed, just released or released.

	Parameters
	----------
		name: string
			The action name.
	"""
	self.name = name


func get_state(which="input"):
	"""
	Returns the current state of the action.

	Parameters
	----------
	which: String {"idle", "fixed", "input"}
		Return state based on process funcion used. If called from `_process`, use `which = "idle"`,
		for `_fixed_process` use `which = "fixed"`, for `_input` use `which = "input"` (default).

	Returns
	-------
		out: int
			0 - IDLE
			1 - PRESSED
			2 - HOLD
			3 - RELEASE
	"""
	var out = null  # output variable
	self.__state_prev[which] = self.__state[which]  # set previous state to current state before getting new state
	self.__state[which] = Input.is_action_pressed(self.name)  # get new state and set it to current state internally

	if not self.__state_prev[which] and self.__state[which]:
		# if there's no previous state but there is a current state then it means PRESSED happened
		out = self.PRESSED
	elif self.__state_prev[which] and self.__state[which]:
		# if there's both previous state and current state then it means it's an on-going press (HOLD)
		out = self.HOLD
	if self.__state_prev[which] and not self.__state[which]:
		# if there's a previous state but not a current state then it means we RLEASED
		out = self.RELEASED
		if which in ["idle", "fixed"]:
			# for "idle" and "fixed", switch manually to IDLE
			self.__state_prev[which] = false
			self.__state[which] = false
	elif not self.__state_prev[which] and not self.__state[which]:
		# in all other cases it's IDLE
		out = self.IDLE

	return out
```

I tested it with:

```gdscript
extends Node


var ActionState = preload("res://action_state.gd")
var action_ui_left = ActionState.new("ui_left")


func _ready():
	self.set_process(true)
	self.set_fixed_process(true)
	self.set_process_input(true)


func _process(dt):
	print("idle ", action_ui_left.get_state("idle"))

func _fixed_process(dt):
	print("fixed ", action_ui_left.get_state("fixed"))


func _input(event):
	print("input ", action_ui_left.get_state())
```

And a sample output:

```
idle 0
input 1
idle 1
idle 2
idle 2
fixed 1
idle 2
idle 2
idle 2
idle 2
idle 2
idle 2
idle 2
fixed 2
idle 2
idle 2
idle 2
idle 2
idle 2
fixed 2
idle 2
idle 2
idle 2
idle 2
idle 2
idle 2
fixed 2
idle 2
idle 2
idle 2
idle 2
idle 2
idle 2
fixed 2
idle 2
idle 2
idle 2
idle 2
idle 2
idle 2
fixed 2
idle 2
input 3
idle 3
idle 0
idle 0
idle 0
input 0
idle 0
idle 0
fixed 3
idle 0
idle 0
idle 0
idle 0
idle 0
idle 0
fixed 0
idle 0
idle 0
idle 0
idle 0
idle 0
idle 0
fixed 0
idle 0
idle 0
idle 0
idle 0
idle 0
idle 0
input 1
fixed 1
idle 1
input 2
idle 2
idle 2
idle 2
idle 2
fixed 2
idle 2
idle 2
```

Which basically confirms it's working as intended, but of course, there's delays, for example `_fixed_process` will be receiving the `RELEASE` output later on, long before `_input` and `_process` get it, I don't know if this is good or bad... I need to put it to the test in actual project, except I don't have one that needs this kind of input checking yet.

**Comment by MarianoGnu** [comment]
what about adding a new node type just for traking input and emmiting signals just_pressed/just_released? we can set a property to select if it uses proccess, or fixed_proccess
i was thinking in adding a shortcut node to track key events, because this could be a solution for editable shortcuts on the editor

**Comment by punto-** [comment]
Where is the _input function is this code?

**Comment by razvanc87** [comment]
@MarianoGnu, don't think that's a good idea. It makes things confusing, what would happen if you want to get the state in both `_process` and `_fixed_process` and what about `_input`? Anyway, the code behind the game engine (the C++ code) probably knows if your GDScript function is called form within `_process` or `_fixed_process` or `_input`, no?

So then, when someone writes `Input.get_state()` in GDScript, the C++ code knows where this was called from and it can return a proper state, like I did with the code above, except you won't have to pass in `"input"`, `"idle"` or `"fixed"`. I mean it all depends if the C++ code knows where functions in GDScript where invoked :d

**Comment by punto-** [comment]
If you _really_ want to run the code on _process instead of _input, why not
push the events you get on _input to a queue and process them on _process
and/or _fixed_process? Or you could try something like this:
http://pastebin.com/xitzHjFM , it's not a good idea to use Input to detect
if a button was pressed, you can lose events like that, and the fact that
you don't have a game that needs this is probably because it's not
something that is usually needed.. Unless you're doing something like
deterministic input, in which case the engine tracking won't help you
either.

**Comment by razvanc87** [comment]
OK, maybe using it inside`_input` isn't a good idea, although it would be weird to miss events since after all you're using the `Input ` class. OK so it might a core game engine implementation thing that I'm not understanding. In that case it should be made clear in the docs that `Input.is_action_pressed` isn't meant to be used inside `_input`. @punto-, about the code, I don't think it works as you intend. The inside `_input ` you're just adding to the stack and in `_process` you keep clearing it, it's almost (not quite, but almost) as having `a.is_pressed` checked inside `_process`. Anyway, I'm actually more interested in having the 4 states available, IDLE, PRESSED, HOLD and RELEASED. Because for example you want to fire something once on PRESSED, but then do something else on HOLD. You can't really do this in a clean way right now, can you? Don't think so...

**Comment by punto-** [comment]
Yes, there is a clean way to do what you describe, you use the _input
method. On _input you get only 1 event when a key is pressed, and 1 event
when a key is released. That's when you run your code that has to run only
once. In between those two events, you don't get any events, that's when
you use Input.is_action_pressed() to see if the key is pressed (what you
call "hold" I assume).
You can use Input.is_action_pressed wherever you like, just don't use it to
detect if a key was just pressed or released. For that you have to use the
events you get in the _input method.

**Comment by razvanc87** [comment]
But this only works for keys, not actions... Isn't that the whole point of actions? To have input from mouse/joystick/touch/keys etc. all abstracted away...

**Comment by punto-** [comment]
Not it also works with actions, it's the same. You can use
event.is_action() to see if the event corresponds with an action.

**Comment by razvanc87** [comment]
Allright, sorry, I mixed up some stuffs actually. Yes I suppose this works for actions :). I was thinking about touch (drag) and mouse motion, which can't be assigned in the input map, my bad. But I still think it's a good idea to have this in the engine since it's such a basic feature and this way the engine doesn't force you to make `_input` just to check if something got pressed and released when you say actually want to know the state in the `_process` and `_fixed_process` functions... it's less confusing for newcomers as well I believe, to just have this `get_action_state` member in the `Input` class... or something on these lines. And as I understand it, make it clear that it isn't meant to be used within `_input`? But that's kind of weird, cause for example `_input` reacts to mouse motion and touch screen drag as well, no? So it would make sense to have some state (pressed, hold, released) for `_input` as well...

**Comment by punto-** [comment]
You can use Input whenever you like, even inside _input. It doesn't make
sense to have a "pressed" state from Input inside _input tho, since _input
is where you get the events. That's how you find out that something was
pressed, because you get an event in _input.

**Comment by razvanc87** [comment]
Well, I didn't necessarily think of having to use `Input` inside `_input`, but more to have it used inside `_process` and `_fixed_process` to get the 4 states. I didn't explain this properly but what I meant is to have the same naming convention for `InputEvent` as well or `InputDefault` (cause I don't know which is the base)... that is `get_action_state()`. Inside `_input` (using `event.get_action_state`) would check if it's already action and what state it is in (pressed/released), for process functions then you'd use `Input.get_action_state` and have it return idle/pressed/hold/released. Does that make sense?...

**Comment by akien-mga** [comment]
What's the status on this proposal guys? :)

**Comment by vnen** [comment]
Based on the discussion, I don't think this is happening. It's too much complicated because Godot have pretty much two frame times (idle and fixed) and you'd have to save state for each and also would have to check the callee to know which state to use. Not to mention that you could have this in nested functions and even separate threads.

As @punto- said, if you need to make it on first press/release, use `_input(event)`. In `_[fixed_]process()` you can assume it's being held. And there's the `is_echo()` already to detect holding.

The only clean way I can think of implementing this would make the user responsible for refreshing the state with a function call (to, say, `Input.refresh_state()`). I'd rather have that as a plugin though.

**Comment by Hinsbart** [comment]
I guess this can be closed since `Input.is_action_just_pressed()` and `Input.is_action_just_released()` have been implemented?

**Comment by vnen** [comment]
Yeah, it was already quite an effort to add those and probably anything else (like tracking each key) would be an overkill.
